packages/main/src/plugin/kubernetes-context-state.ts

```diff
@@ -0,0 +1,295 @@
+﻿/**********************************************************************
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+import type {
+  Informer,
+  KubernetesObject,
+  ListPromise,
+  ObjectCache,
+  V1Deployment,
+  V1Pod,
+} from '@kubernetes/client-node';
+import { AppsV1Api, CoreV1Api, KubeConfig, makeInformer } from '@kubernetes/client-node';
+import type { KubeContext } from './kubernetes-context.js';
+import type { ApiSenderType } from './api.js';
+
+// ContextInternalState stores informers for a kube context
+interface ContextInternalState {
+  podInformer?: Informer<V1Pod> & ObjectCache<V1Pod>;
+  deploymentInformer?: Informer<V1Deployment> & ObjectCache<V1Deployment>;
+}
+
+// ContextState stores information for the user about a kube context: is the cluster reachable, the number
+// of instances of different resources
+export interface ContextState {
+  reachable: boolean;
+  podsCount: number;
+  deploymentsCount: number;
+}
+
+interface CreateInformerOptions<T> {
+  checkReachable?: boolean;
+  onAdd?: (obj: T) => void;
+  onDelete?: (obj: T) => void;
+  onReachable?: (reachable: boolean) => void;
+  timer: NodeJS.Timeout | undefined;
+  backoff: Backoff;
+}
+
+class Backoff {
+  private readonly initial: number;
+  constructor(
+    public value: number,
+    private readonly max: number,
+    private readonly jitter: number,
+  ) {
+    this.initial = value;
+    this.value += this.getJitter();
+  }
+  get() {
+    const current = this.value;
+    if (this.value < this.max) {
+      this.value *= 2;
+      this.value += this.getJitter();
+    }
+    return current;
+  }
+  reset() {
+    this.value = this.initial + this.getJitter();
+  }
+
+  private getJitter(): number {
+    // This is not security-related code
+    return Math.floor(this.jitter * Math.random()); // NOSONAR
```

Comment:
for disabling sonar rules you can do that from the UI of sonar source

but maybe you could just use
```ts
const uintArray = new Uint8Array(1)
const randomInt = crypto.getRandomValues(uintArray)[0]
const randomFloat = randomInt / Math.pow(2, 8)
```

Reply:
> for disabling sonar rules you can do that from the UI of sonar source
> 

OK, I'll check agin the UI. Thanks

> but maybe you could just use
> 
> ```ts
> const uintArray = new Uint8Array(1)
> const randomInt = crypto.getRandomValues(uintArray)[0]
> const randomFloat = randomInt / Math.pow(2, 8)
> ```

As it is really not crypto code, I would prefer to keep it simple